rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,370 @@
+
+### The "connection upgrade"
+
+Borrowed from the "WebSocket" design, the Teleport client can make a connection upgrade through a web API on the
+Teleport Proxy:
+
+```
+ ┌───┐                     ┌─────────────┐                 ┌──────────────┐
+ │tsh│                     │Load Balancer│                 │Teleport Proxy│
+ └─┬─┘                     └───────┬─────┘                 └───────┬──────┘
+   │                               │                               │
+   │ GET /webapi/connectionupgrade │                               │
+   │ Upgrade: alpn                 │                               │
+   ├──────────────────────────────►│                               │
+   │                               │  ┌───────────────┐            │
+   │                               ├──┤TLS Termination│            │
+   │                               │  └───────────────┘            │
+   │                               │                               │
+   │                               │ GET /webapi/connectionupgrade │
+   │                               │ Upgrade: alpn                 │
+   │                               ├──────────────────────────────►│
+   │                               │                               │
+   │                               │ HTTP 101 Switching Protocols  │
+   │                               │◄──────────────────────────────┤
+   │ HTTP 101 Switching Protocols  │                               │
+   │◄──────────────────────────────┤                               │
+   │                               │                               │
+   │ TLS Routing: teleport-mysql   │                               │
+   │◄──────────────────────────────┼──────────────────────────────►│
+   │                               │                               │
+```
+
+This allows the client to tunnel the "original TLS Routing" call through a Layer 7 (HTTP) load balancer.
+
+### Teleport Proxy with self-signed certs
+
+Load balancers usually terminate TLS with certificates signed by publicly trusted CAs. Since Teleport Proxy is no longer
+the public entry point, it is no longer necessary for Teleport Proxy to serve a publicly trusted cert (e.g. using ACME).
+Thus Teleport Proxy should be welcome to use self-signed certs when sitting behind load balancers.
+
+To accommodate that, instead of serving the **Web** certs, the ALPN server on the Proxy will serve the **Host** certs
+when receiving connections from the upgrade flow. And instead of using the SystemCertPool, the Teleport client should
+use the **Host** CA for verifying the Proxy server when dialling TLS Routing inside an upgrade connection. Note that
+this modification does NOT apply to Teleport ALPN protocols that already do mTLS with other certs.
+
+### When to make a connection upgrade
+
+Teleport clients should first perform a test to decide if a connection upgrade is necessary. The test is done by sending a
+TLS handshake with a Teleport custom ALPN to the Proxy server:
```

Comment:
Does this imply that every connection will need to wait for a completely separate TCP and TLS handshake first? That severely hampers the work done lately by @rosstimothy to reduce `tsh` latency as much as possible. Can we figure out some other way to achieve this?

Running normal, direct connections in parallel with the one that will attempt an upgrade might work, for instance - and it would be nice to also keep some short-ish local cache of addresses that are reachable directly or via connection upgrade, so we don't have to always try both.

Reply:
In the current implementation, we do cache the result in tsh profile so only doing this once per login.

The slow part is the upgrade itself, adding 2 round trips (TLS 1.3) per new connection.